paddle/framework/framework.md

```diff
@@ -0,0 +1,205 @@
+# Design Doc: Framework
+
+This design is after the learning of other deep learning systems and numerous group discussions.
+
+## Operation and Operator
+
+At the core of a deep learning system is the neural network.  A neural network is a directed graph of operations, where each operation is an instance of an *operator*, which is a C++ class with a `Run` method derived from the base class `Operator`. 
+
+### `OperationProto` 
+
+Users are supposed to describe neural networks by calling Python.  functions known as *operation creators*, which, in turn call a C++ function `paddle::framework::CreateOperation`.  This C++ function eases the adding of more language bindings.
+
+Python and C++ have different function call syntax, so we define the parameter of `paddle::framework::CreateOperation` a protobuf message `OperationProto`.
+
+### `OperatorProto`
+
+We'd like to generate operator creators automatically from C++ code so to keep Python code always updated.  So we need to describe each C++ class in a protobuf message `OperatorProto`. We also need to fill in an `OperatorProto` message for each operator class and expose these messages to Python function `paddle.framework.create_operation_creators`.  We call this filling and exposing mechanism *operator registration*.
+
+
+## Operators, Layers, Variables, Scope
+
+### Operators as Functions
+
+An operator is intrinsically a function, which has inputs and outputs.  For example, the functional representation of the GEMM operator is
+
+```cpp
+gemm(X, W, scale, act=LeRU) {
+  unactivated = scale * X * W
+  if !act {
+    return unactivated
+  }
+  return unactivated, act(unactivated, cap)
+}
+```
+
+Note that operators might call other operators.  In above example, `gemm` calls `act`.
+
+### Gradient Operators
+
+Each operator has a corresponding gradient operator that defines the gradient computation.
```

Comment:
One forward op is not only corresponding to one gradient operator but gradient operators.

For example, `FcOp` has three inputs. Not all of them needs gradient in practice. It could implement each input corresponding to one gradient operator, so `FcOp` could have three gradient ops.

More generally, Caffe2 and MXNet return an array of ops as one op's gradient ops.